### PR TITLE
fix(42133): undefined[] converted to any[] in js 

### LINKED ISSUE #42133: undefined[] converted to any[] in js
**TypeScript Version:**  4.2.0-dev.20201222

**Search Terms:** undefined any array

**Code**
In .ts everything is ok:

```ts
const a = [1, 2, undefined, 3]

const isUndef = (v: unknown): v is undefined => v === undefined

// OK: undefined[]
const b = a.filter(isUndef)
```

.js:

```js
const a = [1, 2, undefined, 3]

/** @type {(v: unknown) => v is undefined} */
const isUndef = v => v === undefined

// Expected: undefined[]
// Actually: any[]
const b = a.filter(isUndef)
```

**Expected behavior:**
Variable `b` has type `undefined[]`

**Actual behavior:**
Filter method returns `undefined[]`, but variable `b` has `any[]` type

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #42133

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -31385,7 +31385,8 @@ namespace ts {
                     reportImplicitAny(declaration, anyType);
                     return anyType;
                 }
-                else if (isEmptyArrayLiteralType(widened)) {
+
+                if (isEmptyArrayLiteralType(widened)) {
```
Comment (sandersn) [inline src/compiler/checker.ts:31389]: just checking, but...this diff doesn't change anything, right?

Comment (sandersn) [reply src/compiler/checker.ts:31389]: Later: I tested locally that it doesn't.

**Review by sandersn — APPROVED** [review]
Nice distinction between strict-null-checks on and off.

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this
@typescript-bot test this
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this

**Comment by sandersn** [comment]
Since this is a 4.3 bug, let's hold this until after 4.2's RC period is over.

File: src/compiler/checker.ts (hunk 1 above)
Comment (a-tarasyuk) [reply src/compiler/checker.ts]: @sandersn Yes, it doesn't change anything, it seems I just forgot to return the `else` after debugging. I reverted that change back. Sorry for the confusion.

**Pull request merged**